### PR TITLE
Fix uiopacity change when parent node changes.

### PR DESCRIPTION
**Pull request by qiuguohua**
Re: #

### Changelog

https://github.com/cocos/cocos-engine/issues/16514

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: cocos/2d/components/ui-opacity.ts [hunk 1]
```diff
@@ -133,8 +134,32 @@ export class UIOpacity extends Component {
...
+        const uiOp = node.getComponent<UIOpacity>(UIOpacity);
+        if (uiOp) {
+            UIOpacity.setEntityLocalOpacityDirtyRecursively(node, true, 1, false);
+        }
+    }
+
     public onEnable (): void {
-        // If the ancestor node has a uiopacity component, it will be initialized when initializing 
+        this.node.on(NodeEventType.CHILD_ADDED, this._parentAdded, this);
+        this.node.on(NodeEventType.CHILD_REMOVED, this._parentRemoved, this);
```
Comment (zhefengzhang) [inline cocos/2d/components/ui-opacity.ts:161]: Need to turn off the listener in the onDisable callback

Comment (qiuguohua) [reply cocos/2d/components/ui-opacity.ts:161]: OK

File: cocos/2d/components/ui-opacity.ts [hunk 2]
```diff
@@ -133,8 +134,32 @@ export class UIOpacity extends Component {
     @serializable
     protected _opacity = 255;
 
+    protected _parentAdded (node: Node): void {
+        const parent: Node |null = node.getParent();
+        if (parent == null) {
+            return;
+        }
```
Comment (minggo) [inline cocos/2d/components/ui-opacity.ts:141]: Why should check it as it is a callback of CHILD_ADDED.

File: cocos/2d/components/ui-opacity.ts [hunk 3]
```diff
@@ -133,8 +134,32 @@ export class UIOpacity extends Component {
...
+        const render = parent._uiProps.uiComp as UIRenderer;
+        if (render && render.color) {
+            return;
```
Comment (minggo) [inline cocos/2d/components/ui-opacity.ts:144]: Why return here?

Comment (qiuguohua) [reply cocos/2d/components/ui-opacity.ts:144]: If it is a renderable element, the child nodes do not need to respond to changes.

File: cocos/2d/components/ui-opacity.ts (hunk 2 above)
Comment (qiuguohua) [reply cocos/2d/components/ui-opacity.ts:141]: Because getParent may be null

Comment (minggo) [reply cocos/2d/components/ui-opacity.ts:141]: It is called when a child is added, then why parent will be null?

**Comment by qiuguohua** [comment]
Fixed in another pr ：https://github.com/cocos/cocos-engine/pull/16828

**Pull request closed without merging**